Fix vlan issues

LKomaryanskiy opened the pull request:
This PR fixes several problems:
1. Removing VLAN devices via "ip link del" command. Previously it led to dumping trace and system hanging;
2. Locking for rdev_list that led to system hanging during VLAN (creation new device add routes etc) operation;
3. System hanging because of writing to inoperable chains when pinging VLAN device after device was deleted;
4. Memory leak in VLAN device unregister.

LKomaryanskiy committed:
net: renesas: rswitch: fix locking in napi_poll

Using mutexes in functions such as rswitch_poll etc leads
to kernel hang up. Changing mutexes to rw_locks fixes this
issue.

Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>

LKomaryanskiy committed:
net: renesas: rswitch: fix null pointer dereference on VLAN remove

For VLAN devices, kernel constructs ndev and fills needed structures
such as dev.parent, but for proper chain mapping R-Switch driver requires
real device parent. So we need to save pointer to ndev->dev.parent and
restore it for proper kernel deinit ndev.

Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>

LKomaryanskiy committed:
net: renesas: rswitch: fix improper removing routes for VLANs

Currently, rswitch_fib_wq is invoked after vlan_dev_unregister
when VLAN rdev is already deleted from rdev list. In this case
forwarding rules are not removed from MFWD l3 table which leads
to an issue when pinging VLAN interface after removing causes
to write to unprocessable GWCA chain and as a result system
hangs up. The patch fixes the issue by explicitly removing
routes in VLAN dev unregister function.

Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>

LKomaryanskiy committed:
net: renesas: rswitch: add netif_napi_del in VLAN unregister

After unregistering VLAN, napi should be properly deleted not only
disabled.

Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>

firscity commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -4121,8 +4148,12 @@ static void vlan_dev_unregister(struct net_device *dev)
+	cleanup_all_routes(rdev);
 
 	list_del(&rdev->list);
+	dev->dev.parent = rdev->vlan_parent;
```
We probably also need `kfree(rdev)` on device unregister, subsequent vlan device register/unregister will lead to memory leak.

LKomaryanskiy committed:
net: renesas: rswitch: fix memory leak with rdev in vlan unregister

R-Switch device for VLANs are allocated via kzalloc in vlan register
netdev event, so free memory in VLAN unregister to avoid memory
leak.

Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>

firscity reviewed (approved):
`Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>`
